add tutorial on Zero Knowledge Proofs with Noir for Rootstock (#433)
rsksmart/devportal: docs/04-resources/04-tutorials/zk-proofs-with-noir.md

---
sidebar_label: Zero-Knowledge Proofs with Noir
sidebar_position: 4
title: Zero-Knowledge Proofs on Rootstock with Noir
description: "Building private decentralized applications (dApps) on Rootstock is key to ensuring user confidentiality and security. Zero-Knowledge Proofs (ZK proofs) and Noir form a powerful combination, blending advanced cryptography with a developer-friendly language to create privacy-focused solutions."
tags: [zk, noir, zero-knowledge, privacy, tutorial, solidity, dapp, rootstock]
---

> # Zero-Knowledge Proofs on Rootstock with Noir

Reviewer: ```suggestion
```

Author: Alright, noted. This will be corrected

> Zero-knowledge proofs let a user prove they know something (a secret code, a credential, ownership) without ever revealing the secret itself. On Rootstock — the Bitcoin-secured, EVM-compatible smart contract chain — this unlocks real privacy while keeping Bitcoin-level security.

Reviewer: ```suggestion
Zero-knowledge proofs let a user prove they know something (a secret code, a credential, ownership) without ever revealing the secret itself. On Rootstock — the Bitcoin-secured, EVM-compatible smart contract chain — this unlocks real privacy while preserving Bitcoin-level security.
```

> This hands-on tutorial teaches you how to use **Noir** (a developer-friendly ZK DSL-Domain Specific Language) to build a **Secret NFT Club**: users get an exclusive membership only by proving they know the secret password — the password never appears on-chain or in the browser console.

Reviewer: Add link to Noir

- The proof is verified on-chain using zero-knowledge cryptography
- Members are able to join the club upon successful verification
- The password never appears in transactions, logs, or browser console

**Privacy guarantee:** Even if someone inspects all blockchain data, they cannot determine the secret password.

## Prerequisites

- Node.js ≥ 18
- [Rust](https://rust-lang.org/tools/install/) (for Noir toolchain)
- MetaMask wallet with tRBTC on Rootstock Testnet ([Get tRBTC from Faucet](https://faucet.rootstock.io/))
- Basic knowledge of Solidity and React/Next.js

:::warning[Note]

🚨 Windows Users: Noir (nargo, bb) isn’t natively supported on Windows. Please install and run Noir inside WSL (Windows Subsystem for Linux) using Ubuntu 24.04.. 🚨

:::

## Part 1: Setup & Circuit Development

### Step 1: Install Noir (Nargo CLI)

We'll use `nargo version = 1.0.0-beta.3`.

```bash
curl -L https://raw.githubusercontent.com/noir-lang/noirup/refs/heads/main/install | bash
noirup -v 1.0.0-beta.3
```

Verify installation:

```bash
nargo --version
# Should output: nargo version = 1.0.0-beta.3
```

### Step 2: Install Barretenberg Backend

> Barretenberg is the proving backend that generates and verifies zero-knowledge proofs.

Reviewer: Add link to Barretenberg? And why it is needed

```bash
curl -L https://raw.githubusercontent.com/AztecProtocol/aztec-packages/refs/heads/master/barretenberg/bbup/install | bash
bbup -v 0.82.2
```

Verify:

```bash

> bb --version

Reviewer: ```
owanate@Owanates-MacBook-Pro ~ % curl -L https://raw.githubusercontent.com/AztecProtocol/aztec-packages/refs/heads/master/barretenberg/bbup/install | bash
bbup -v 0.82.2
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1896  100  1896    0     0   9161      0 --:--:-- --:--:-- --:--:--  9203
Downloading bbup...
✓ Successfully installed bbup
Please run 'source ~/.bashrc' or restart your terminal to use bbup
zsh: command not found: bbup
owanate@Owanates-MacBook-Pro ~ % bb --version
zsh: command not found: bb
owanate@Owanates-MacBook-Pro ~ % 

```

Output:

```
owanate@Owanates-MacBook-Pro ~ % bb --version
zsh: command not found: bb
```

Reviewer: Ran `source ~/.zshrc`, it resolved, but no version number outputted

```
owanate@Owanates-MacBook-Pro ~ % bb --version  
zsh: command not found: bb
owanate@Owanates-MacBook-Pro ~ % bbup --version
owanate@Owanates-MacBook-Pro ~ % 

```

Ensure to include this in a troubleshooting section

Reviewer: Fixed by re-running command in .zshrc shell

```
owanate@Owanates-MacBook-Pro ~ % curl -L https://raw.githubusercontent.com/AztecProtocol/aztec-packages/refs/heads/master/barretenberg/bbup/install | bash bbup -v 0.82.2
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0Installing to /Users/owanate/.bb
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1896  100  1896    0     0   9677      0 --:--:-- --:--:-- --:--:--  9723
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 10.4M  100 10.4M    0     0  13.7M      0 --:--:-- --:--:-- --:--:-- 49.7M
✓ Installed barretenberg to /Users/owanate/.bb
owanate@Owanates-MacBook-Pro ~ % bb --version
v0.82.2
owanate@Owanates-MacBook-Pro ~ % 
```

# Should output: v0.82.2
```

### Step 3: Create the ZK Circuit

Create a new Noir project:

```bash
nargo new secret_club
cd secret_club
```

Replace `src/main.nr` with this circuit:

```rust
use std::hash::pedersen_hash;

fn main(secret: Field, public_hash: pub Field) {
    let computed_hash = pedersen_hash([secret]);
    assert(computed_hash == public_hash);
}
```

**What this does:**

- Takes a `secret` (private input - never revealed)
- Takes a `public_hash` (public input - visible to everyone)

> - Computes Pedersen hash of the secret

Reviewer: Please link to meaning of these terms or introduce them in a glossary section:

`Pedersen`

```

#### Now let's compute the Pedersen Hash

To compute the pedersen hash, we'll slightly modify our `main.nr`
We're adding a println to print the perderson hash and we're writing a test to output this hash to the console.

```rust
use std::hash::pedersen_hash;

fn main(secret: Field, public_hash: pub Field) {
    let computed_hash = pedersen_hash([secret]);

    println(computed_hash); // we added this line to print the perderson hash

    assert(computed_hash == public_hash);
}

#[test]
fn test_main() {
    main(
        0x04e94fe643fe9000c83dd91f0be27855aa2cd791a3dfc1e05775749e89f4693e,
        0x3, // this is just a placeholder for the public hash which will cause the test to fail, but we will get the perderson hash logged to the console
    );
}
```

Then in your terminal, run the command

```bash
nargo test --show-output
```

Look for the **test_main stdout** in the output - this is your Pedersen hash!

Example output:

```
--- test_main stdout ---
0x297fad8a9bc7f877e7ae8ab582a32a16ec2d11cc57cd77ecab97d2c775fa29e8
------------------------
```

**Save this hash!** You'll need it for:

- Smart contract deployment
- Frontend configuration
- Testing

> Compile and Execute: these generate the `secret_club.json` and `secret_club.gz` files respectively, which will be used as we proceed

Reviewer: ```suggestion
Compile and Execute: The compile and execute commands generates the `secret_club.json` and `secret_club.gz` files respectively, which will be used as we proceed
```

```bash
nargo compile

> nargo execute

Reviewer: ```
owanate@Owanates-MacBook-Pro secret_club % nargo execute
Cannot find input file '/Users/owanate/Documents/Learning/secret_club/Prover.toml'
```

Reviewer: @michojekunle I get this error when I run nargo execute

Author: Oh my, my oversight, I skipped an important step in the guide, I'll add that now.

Reviewer: Thank you

Reviewer: Output:

```
nargo execute
0x297fad8a9bc7f877e7ae8ab582a32a16ec2d11cc57cd77ecab97d2c775fa29e8
[secret_club] Circuit witness successfully solved
[secret_club] Witness saved to target/secret_club.gz
```

------------------------
```

**Save this hash!** You'll need it for:

- Smart contract deployment
- Frontend configuration
- Testing

Before we can proceed to run the `nargo execute` command, we need to generate a `Prover.toml` file.
This file holds the witness values (i.e. the `secret` and the `public_hash`).
To generate it, we start by running:

```bash
nargo check
```

Running `nargo check` creates a new `Prover.toml` file, prefilled based on the inputs defined in the main function of our `main.nr` circuit:

```toml
public_hash = ""
secret = ""
```

Now we can fill in these fields with our actual witness values — the hashed `secret` (for example, the SHA-256 hash of 'supersecret2025') and the `public_hash` (the corresponding Pedersen hash):

```toml
public_hash = "0x297fad8a9bc7f877e7ae8ab582a32a16ec2d11cc57cd77ecab97d2c775fa29e8"
secret = "0x04e94fe643fe9000c83dd91f0be27855aa2cd791a3dfc1e05775749e89f4693e"
```

Once the `Prover.toml` file is filled, you can proceed to compile and execute the circuit:

```bash
nargo compile
nargo execute
```

These commands generate the `secret_club.json` and `secret_club.gz` files, which we will use moving forward.

:::warning[Note]

> 🚨🚨 Always delete the files in the target folder when you change your circuit or inputs to ensure a clean setup. Whenever the circuit changes, you must also regenerate and replace the verifier smart contract in your Solidity project. 🚨🚨

Reviewer: ```suggestion
🚨🚨 Always delete the files in the `target` folder when you change your circuit or inputs to ensure a clean setup. Whenever the circuit changes, you must also regenerate and replace the verifier smart contract in your Solidity project. 🚨🚨
```

Author: Alright, I'll add this

:::

### Step 5: Generate the Solidity Verifier

Modern Noir uses Barretenberg to generate the Solidity verifier:

```bash
# Generate verification key

> bb write_vk --oracle_hash keccak -b ./target/secret_club.json -o ./target

Reviewer: Do I run both commands? Please specify on the guide or separate the code blocks for each

Output:

```
bb write_vk --oracle_hash keccak -b ./target/secret_club.json -o ./target
Scheme is: ultra_honk
Length is too large
owanate@Owanates-MacBook-Pro secret_club % 
```

# Generate Solidity verifier contract

> bb write_solidity_verifier -k ./target/vk -o ./target/Verifier.sol

Reviewer: Output:

```
bb write_solidity_verifier -k ./target/vk -o ./target/Verifier.sol
Scheme is: ultra_honk
Unable to open file: ./target/vk
owanate@Owanates-MacBook-Pro secret_club % 

```

Reviewer: I probably missed something, do I need to clear those files in the target folder before running these two commands above to generate verifier.sol?

Author: yes, that's also noted in the guide

```

> This creates `Verifier.sol` in the `./target/Verifier.sol`. The vk is embedded into this contract, enabling Rootsock to check proofs generated for your circuit.

Reviewer: Couldn't create verifier.sol file in target folder

Author: Deleting the target folder and re-running the commands should fix this, or does it not? So I can try to debug from my end

Reviewer: This is the error I get after deleting the target folder

```
secret_club % bb write_vk --oracle_hash keccak -b ./target/secret_club.json -o ./target
Scheme is: ultra_honk
jq: error: Could not open file ./target/secret_club.json: No such file or directory
gunzip: (stdin): unexpected end of file
Input is not large enough
owanate@Owanates-MacBook-Pro secret_club % 
```

Author: Hi @owans  After deleting the target folder, you’d have to compile and execute again by running ‘nargo compile’ and ‘nargo execute’ before running the next commands

Reviewer: Ok, noted

Reviewer: Here's the output:

```
nargo compile
owanate@Owanates-MacBook-Pro secret_club % nargo execute
0x297fad8a9bc7f877e7ae8ab582a32a16ec2d11cc57cd77ecab97d2c775fa29e8
[secret_club] Circuit witness successfully solved
[secret_club] Witness saved to target/secret_club.gz
owanate@Owanates-MacBook-Pro secret_club % bb write_vk --oracle_hash keccak -b ./target/secret_club.json -o ./target
Scheme is: ultra_honk
Length is too large
owanate@Owanates-MacBook-Pro secret_club % bb write_solidity_verifier -k ./target/vk -o ./target/Verifier.sol
Scheme is: ultra_honk
Unable to open file: ./target/vk
owanate@Owanates-MacBook-Pro secret_club % 
```

Reviewer: Done but still having errors, see the the terminal output:

```
nargo compile
owanate@Owanates-MacBook-Pro secret_club % nargo execute
0x297fad8a9bc7f877e7ae8ab582a32a16ec2d11cc57cd77ecab97d2c775fa29e8
[secret_club] Circuit witness successfully solved
[secret_club] Witness saved to target/secret_club.gz
owanate@Owanates-MacBook-Pro secret_club % bb write_vk --oracle_hash keccak -b ./target/secret_club.json -o ./target
Scheme is: ultra_honk
Length is too large
owanate@Owanates-MacBook-Pro secret_club % bb write_solidity_verifier -k ./target/vk -o ./target/Verifier.sol
Scheme is: ultra_honk
Unable to open file: ./target/vk
owanate@Owanates-MacBook-Pro secret_club % 
```

Author: Alright Please can you check that the versions of bb and nargo are as specified in the guide.

@owans

Reviewer: Hi @michojekunle 

Here it is:

```
secret_club % bb --version
v0.82.2
```

Also ran nargo command:
```
 nargo --version
nargo version = 1.0.0-beta.15
noirc version = 1.0.0-beta.15+83245db91dcf63420ef4bcbbd85b98f397fee663
(git version hash: 83245db91dcf63420ef4bcbbd85b98f397fee663, is dirty: false)
owanate@Owanates-MacBook-Pro secret_club % 
```

Author: Hi @owans,

Please can you confirm again the versions you’re using — is bb v0.82.2 and nargo v1.0.0-beta.3?

<img width="599" height="119" alt="Screenshot 2025-12-16 at 15 34 35" src="https://github.com/user-attachments/assets/2b181366-117e-498a-901b-09aad549cd2b" />

I just re-scaffolded the app from scratch following the guide and tried all the steps up to the point where you encountered the error, to see if I could reproduce it.

However, I wasn’t able to reproduce the error on my end. I still think it might be related to the versions of the tools (bb and nargo).

You could also try deleting what you have so far and following the guide again from the beginning to see if that fixes it.

## Part 2: Smart Contract Development

### Step 6: Create the Secret NFT Club Contract

#### Setup Hardhat

Install dependencies:

```bash

> mkdir smart-contracts

Reviewer: Specify if command should be ran in the same secret_club folder

Do you want to run it now? (Y/n) · true

npm install --save-dev "hardhat@^3.0.16" "@nomicfoundation/hardhat-toolbox-mocha-ethers@^3.0.1" "@nomicfoundation/hardhat-ethers@^4.0.2" "@nomicfoundation/hardhat-ignition@^3.0.5" "@types/chai@^4.2.0" "@types/chai-as-promised@^8.0.1" "@types/mocha@>=10.0.10" "@types/node@^22.8.5" "chai@^5.1.2" "ethers@^6.14.0" "forge-std@foundry-rs/forge-std#v1.9.4" "mocha@^11.0.0" "typescript@~5.8.0"
npm warn deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
npm warn deprecated glob@7.1.7: Glob versions prior to v9 are no longer supported

added 275 packages, and audited 276 packages in 57s

71 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
✨ Dependencies installed ✨
Give Hardhat a star on Github if you're enjoying it! ⭐️✨

     https://github.com/NomicFoundation/hardhat
➜  smart-contracts git:(main) ✗
```

> Create `contracts/SecretNFTClub.sol`:

Reviewer: The user can delete existing files like counter. sol and counter.t.sol and also the temp file in the deploy folder

        publicInputs[0] = secretHash;

        // Verify the zero-knowledge proof
        if (!verifier.verify(proof, publicInputs)) revert InvalidProof();

        // Proof verified! Grant membership
        uint256 tokenId = _nextTokenId++;
        hasJoined[msg.sender] = true;
        memberTokenId[msg.sender] = tokenId;

        emit MemberJoined(msg.sender, tokenId);
    }

    function isMember(address account) external view returns (bool) {
        return hasJoined[account];
    }

    function totalMembers() external view returns (uint256) {
        return _nextTokenId;
    }
}
```

Make sure to also copy the the `Verifier` contract from `./target/Verifier.sol` into your smart-contracts directory in a new file `contracts/Verifier.sol`. This contract will also be deployed and will be used on our `SecretNFTClub` contract to verify a proof.

**Design decisions:**

- Simple mapping-based membership (more gas-efficient than ERC721)
- Immutable verifier and hash (gas optimization + security)
- Custom errors (saves gas over require strings)
- Events for off-chain tracking

## Part 3: Deployment

### Step 7: Deploy to Rootstock Testnet

**Rootstock Testnet Details:**

> 

Reviewer: Specify these details can be configured on their software wallet or refer them to this page: https://dev.rootstock.io/dev-tools/wallets/metamask/

| Parameter      | Value                                       |
| -------------- | ------------------------------------------- |
| RPC URL        | `https://public-node.testnet.rsk.co`        |
| Chain ID       | `31`                                        |
| Currency       | tRBTC                                       |
| Block Explorer | `https://rootstock-testnet.blockscout.com/` |
| Faucet         | `https://faucet.rootstock.io`               |

Configure `hardhat.config.ts`:

```javascript

> import hardhatToolboxMochaEthersPlugin from "@nomicfoundation/hardhat-toolbox-mocha-ethers";

Reviewer: Add dotenv to the installation requirements for the Hardhat project (smart_contracts)

<img width="829" height="111" alt="Image" src="https://github.com/user-attachments/assets/558f327f-86cf-4476-acee-ba0d43aadc65" />

  const Club = await ethers.getContractFactory("SecretNFTClub");
  const club = await Club.deploy(SECRET_HASH, verifierAddress);
  await club.waitForDeployment();

  const clubAddress = await club.getAddress();
  console.log("✅ SecretNFTClub deployed:", clubAddress);

  // Summary
  console.log("\n" + "=".repeat(50));
  console.log("📋 DEPLOYMENT SUMMARY");
  console.log("=".repeat(50));
  console.log("Verifier:    ", verifierAddress);
  console.log("Club:        ", clubAddress);
  console.log("Secret Hash: ", SECRET_HASH);
  console.log("Network:     ", "Rootstock Testnet");
  console.log(
    "Explorer:    ",
    `https://explorer.testnet.rootstock.io/address/${clubAddress}`
  );
  console.log("=".repeat(50));

  // Save addresses for frontend
  fs.writeFileSync(
    "deployment.json",
    JSON.stringify(
      {
        verifier: verifierAddress,
        club: clubAddress,
        secretHash: SECRET_HASH,
        network: "rootstock",
      },
      null,
      2
    )
  );

  console.log("\n✅ Addresses saved to deployment.json");
}

main()
  .then(() => process.exit(0))
  .catch((error) => {
    console.error(error);
    process.exit(1);
  });
```

Deploy:

```bash

> npx hardhat run scripts/deploy.js --build-profile production --network rootstock

Reviewer: This command is misleading as it specifies production instead of testnet, add an explanation of the parameters and why its like this

```

Expected output:

```
🚀 Deploying to Rootstock Testnet...

📝 Deploying HonkVerifier...
✅ UltraVerifier deployed: 0x1234...

📝 Deploying SecretNFTClub...

> ✅ SecretNFTClub deployed: 0x5678...

Reviewer: Provide a brief exaplanation of the deployment summary and what it does:

```
npx hardhat run scripts/deploy.js --build-profile production --network rootstock
[dotenv@17.2.3] injecting env (2) from .env -- tip: 🔐 prevent committing .env to code: https://dotenvx.com/precommit
Downloading solc 0.8.28
Downloading solc 0.8.28 (WASM build)

🚀 Deploying to Rootstock Testnet...

📝 Deploying HonkVerifier...
✅ HonkVerifier deployed: 0x25D1C679461023079679C771c336655B7e637682

📝 Deploying SecretNFTClub...
✅ SecretNFTClub deployed: 0x2d1690F8CcD7e1490D32a9025804a9F6B5d8AD54

==================================================
📋 DEPLOYMENT SUMMARY
==================================================
Verifier:     0x25D1C679461023079679C771c336655B7e637682
Club:         0x2d1690F8CcD7e1490D32a9025804a9F6B5d8AD54
Secret Hash:  0x297fad8a9bc7f877e7ae8ab582a32a16ec2d11cc57cd77ecab97d2c775fa29e8
Network:      Rootstock Testnet
Explorer:     https://explorer.testnet.rootstock.io/address/0x2d1690F8CcD7e1490D32a9025804a9F6B5d8AD54
==================================================

✅ Addresses saved to deployment.json
owanate@Owanates-MacBook-Pro smart-contracts % 
```

```

## Part 4: Frontend Integration

### Step 8: Setup Frontend Project

> Create a new Vite + React project:

Reviewer: Specify if root folder secret_club or other

```bash
npm create vite@latest secret-club-frontend -- --template react
cd secret-club-frontend
```

Install dependencies:

```bash

> npm install @noir-lang/noir_js@1.0.0-beta.3 @aztec/bb.js@0.82.0 ethers

Reviewer: This automatically ran the server on `http://localhost:5173`, you might want to inform the user to split terminal or end CTRL C before going to the next step and starting with npm run dev later.

  },
  status: {
    textAlign: "center",
    marginTop: "20px",
    fontStyle: "italic",
    color: "#666",
    minHeight: "24px",
  },
  info: {
    marginTop: "30px",
    padding: "20px",
    background: "#f7f7f7",
    color: "#000",
    borderRadius: "8px",
    fontSize: "14px",
  },
  list: {
    marginTop: "10px",
    paddingLeft: "20px",
    lineHeight: "1.8",
  },
};
```

Then update your `App.jsx` file to include `JoinClub.jsx`

```jsx
import "./App.css";
import JoinClub from "./JoinClub";

function App() {
  return (
    <>
      <JoinClub />
    </>
  );
}

export default App;
```

### Step 10: Run the Application

Start the development server:

```bash
npm run dev
```

Open http://localhost:5173 in your browser.

<Video url="/video/resources/tutorials/zk-proofs-with-noir/01-secret-nft-club-dapp-demo.mp4" />

**Testing the flow:**

1. Click "Connect Wallet"
2. Click "Join Club (ZK Proof)"
3. Enter password: `supersecret2025`
4. Wait for proof generation (~10-20 seconds first time)
5. Confirm MetaMask transaction
6. Success! You're now a member

![Secret NFT Club - Valid member](/img/resources/tutorials/zk-proofs-with-noir/01-secret-nft-club-valid-member.png)

> ## Understanding the System

Reviewer: ```suggestion
## Understanding the Architecture
```

- Skip error handling

### Preventing Common Attacks

**Front-running:** Add nonce or msg.sender binding to circuit  
**Replay attacks:** Include timestamp or chain ID in proof  
**DoS attacks:** Implement rate limiting and gas limits  
**Secret leakage:** Never log the password or intermediate values

## Troubleshooting Guide

### "Invalid proof" error

**Symptoms:** Transaction reverts with InvalidProof error

**Causes:**

- Wrong password entered
- Hash mismatch between circuit and contract
- Corrupt proof data

**Solutions:**

1. Verify SECRET_HASH in deployment.json matches circuit
2. Check password spelling
3. Regenerate proof
4. Verify circuit compilation

### Proof generation freezes browser

**Symptoms:** UI becomes unresponsive during proof generation

**Cause:** WASM computation blocking main thread

**Solution:** Use Web Workers (see Advanced section below)

### "Out of gas" error

**Symptoms:** Transaction fails with out of gas

**Cause:** Gas limit too low for verification

**Solution:** Increase gas limit manually

```javascript
const tx = await club.join(proof.proof, {
  gasLimit: 5000000,
});
```

## Advanced Topics

### Adding More Complex Proofs

> Extend the circuit to prove multiple conditions:

Reviewer: Earlier, you mentioned changing circuit requires deleting the target folder? If so, you can specify to reun the nargo compile on root folder or let the user know its not required

```rust
use std::hash::pedersen_hash;

fn main(
    secret: Field,
    age: Field,
    public_hash: pub Field,
    min_age: pub Field
) {
    // Verify secret knowledge
    let computed_hash = pedersen_hash([secret]);
    assert(computed_hash == public_hash);

    // Verify age requirement
    assert(age >= min_age);
}
```

This proves: "I know the secret AND I'm over 18" without revealing either value.

### Batch Verification

> To verify multiple proofs efficiently:

Reviewer: What file is this related to in the smart contracts folder?

```solidity
function batchJoin(
    bytes[] calldata proofs,
    address[] calldata members
) external onlyOwner {
    require(proofs.length == members.length, "Length mismatch");

    bytes32[] memory publicInputs = new bytes32[](1);
    publicInputs[0] = secretHash;

    for (uint i = 0; i < proofs.length; i++) {
        require(!hasJoined[members[i]], "Already member");
        require(verifier.verify(proofs[i], publicInputs), "Invalid proof");

        hasJoined[members[i]] = true;
        memberTokenId[members[i]] = _nextTokenId++;

        emit MemberJoined(members[i], _nextTokenId - 1);
    }
}
```

### Testing Your Circuit

> 

Reviewer: Remind user to change directory to run this commands

---
sidebar_lable: Zero Knowledge Proofs on Rootstock with Noir
sidebar_position: 4
title: Zero Knowledge Proofs on Rootstock with Noir
description: 'Building private decentralized applications (dApps) on Rootstock is key to ensuring user confidentiality and security. Zero-Knowledge Proofs (ZK proofs) and Noir form a powerful combination, blending advanced cryptography with a developer-friendly language to create privacy-focused solutions.'
tags: [rsk, rootstock, tutorials, resources, zk, zer-knowledge-proofs, smart contracts, dapps, noir]
---

> ### Introduction

Reviewer: ```suggestion
```

> Zero-Knowledge Proofs (ZK proofs) represent a groundbreaking cryptographic technique that enhances privacy and security in blockchain applications. By allowing one party to prove the validity of a statement without revealing underlying data, ZK proofs enable confidential transactions, private computations, and verifiable claims on public ledgers like Rootstock. Rootstock, as the pioneering Bitcoin Layer 2 (L2) platform, combines Bitcoin's unmatched security through merged mining with Ethereum-compatible smart contracts via the EVM. This makes it an ideal environment for deploying ZK-enhanced dApps, leveraging over 60% of Bitcoin's hash power for robust protection.

Reviewer: ```suggestion
Zero-Knowledge Proofs (ZK proofs) is a cryptographic technique that enhances privacy and security in blockchain applications. By allowing one party to prove the validity of a statement without revealing underlying data, ZK proofs enable confidential transactions, private computations, and verifiable claims on public ledgers such as Rootstock. Rootstock combines Bitcoin's security through merged mining with Ethereum-compatible smart contracts via the EVM. This makes it an ideal environment for deploying ZK-enhanced dApps, leveraging over 60% of Bitcoin's hash power for enhanced protection.
```

Reviewer: Add link to ZK Proof: https://en.wikipedia.org/wiki/Zero-knowledge_proof

Author: Alright, thank you. This link would be added

Rootstock is a blockchain platform that extends the capabilities of the Bitcoin network by incorporating smart contract functionality. Built to be EVM (Ethereum Virtual Machine) compatible, Rootstock enables developers to deploy and execute smart contracts using the same programming languages and tools as Ethereum.

> In this guide, we'll critically explore ZK proofs, introduce Noir—a domain-specific language (DSL) for writing ZK circuits—and demonstrate how to integrate them on Rootstock for improved privacy and security. We'll culminate with a hands-on demo: building a "Secret NFT Club" dApp where users can join only by proving knowledge of a secret code, without revealing it. This showcases real-world applications like exclusive access controls in NFTs, while addressing potential challenges such as computational overhead.

Reviewer: ```suggestion
In this guide, we'll explore ZK proofs, introduce Noir—a domain-specific language (DSL) for writing ZK circuits—and demonstrate how to integrate them on Rootstock for enhanced privacy and security. We will build a "Secret NFT Club" dApp where users can join only by proving knowledge of a secret code, without revealing it. This showcases real-world applications like exclusive access controls in NFTs, while addressing potential challenges such as computational overhead.
```

Reviewer: Simplify the term computational overhead
